### DISCUSSION TITLE
Our Windows & macOS nightly builds, and Linux AppImage now contain SQLite 3.41.1

### DISCUSSION THREAD

**Discussion by justinclift** [sqlitebrowser/sqlitebrowser#3297] [Show and tell]
* Windows nightly builds: https://nightlies.sqlitebrowser.org/latest/
* Linux AppImage: https://github.com/sqlitebrowser/sqlitebrowser/releases/tag/continuous

Haven't updated the macOS nightly builds yet, as that'll take a bit more free time than I have at the moment.  Probably get it done over the weekend.

**Comment by justinclift** [comment]
@deepsidhu1313 A new SQLite release is out. :smile:

**Reply by scottfurry** [reply]
Given how frequently SQLite is released, this has the potential to become a *whack-a-mole* scenario. Just my $/£/€0.02.

**Reply by justinclift** [reply]
Yep, it probably will be.

**Reply by justinclift** [reply]
On that note... yeah, we might want to automate the bits that shouldn't be too hard.

Can think of how to do that for the AppImage and *maybe* the macOS packaging.  The Windows nightly one is probably do-able too.

Will think over that later.  Probably when the next SQLite release comes out. :wink:

**Reply by deepsidhu1313** [reply]
this is the recipe being used in launchpad to build sqlite packages
```
git-build-recipe format 0.4 deb-version 3.41.0-0-{revno}~{revtime}~{time}
[lp:~linuxgndu/sqlite-recipe/+git/sqlite-recipe](https://code.launchpad.net/+code/~linuxgndu/sqlite-recipe/+git/sqlite-recipe) branch-3.41
nest-part packaging [lp:sqlite-recipe](https://code.launchpad.net/+code/sqlite-recipe) debian
nest-part www [lp:sqlite-recipe](https://code.launchpad.net/+code/sqlite-recipe) www
```

Only thing for now i have to change is `branch-3.41`. Is there a new branch on their project that we want to use ?

**Reply by deepsidhu1313** [reply]
link to recipe https://code.launchpad.net/~linuxgndu/+recipe/sqlite-recipe-daily

**Reply by justinclift** [reply]
~Nah, same branch ('master') to build from for now, at least for the nightlies.~

Oh, I misunderstood.  You're building from the sqlite repo rather than downloading the tarball.  With that, I'm not sure.

For the Linux AppImage builds, it downloads the sqlite tarball (.tar.gz) each time.  I wonder if we could put the SQLite download url into (maybe?) some environment variable that the AppImage build script can use?  We could have a different script automatically check for new SQLite release (each day?) then automatically updates that environment variable.

That *might* work reasonably reliably.

**Reply by deepsidhu1313** [reply]
if we are using debian/ubuntu runner to build AppImage, we can add our sqlite ppa to install latest version. Then build sqlitebrowser and Appimage from it.

**Reply by justinclift** [reply]
Oh, that sounds like a good idea.

Looking at the appimage.yml file that says how to build things, it's using Ubuntu 18.04:

https://github.com/sqlitebrowser/sqlitebrowser/blob/9426509ac8a8422d8b164c2acfb6a849ea37e306/.github/workflows/appimage.yml#L21

Then it's doing standard Ubuntu `apt install` stuff to install dependencies:

https://github.com/sqlitebrowser/sqlitebrowser/blob/9426509ac8a8422d8b164c2acfb6a849ea37e306/.github/workflows/appimage.yml#L29

Then it's manually compiling SQLite, using the same options as our macOS nightly builds:

https://github.com/sqlitebrowser/sqlitebrowser/blob/9426509ac8a8422d8b164c2acfb6a849ea37e306/.github/workflows/appimage.yml#L44

With all that done, it then creates an AppImage from it:

https://github.com/sqlitebrowser/sqlitebrowser/blob/9426509ac8a8422d8b164c2acfb6a849ea37e306/.github/workflows/appimage.yml#L49-L69

So, if we update that to use our SQLite PPA instead then we'd only have the one Linux SQLite piece to update each time. :smile:

**Comment by justinclift** [comment]
Just (manually) updated our macOS nightly builds to include SQLite 3.41.1.  Next time around, lets think about whether we can start scripting this or something. :smile:

**Comment by lucydodo** [comment]
Mmmm.. It's abstract, but here's my idea.

`curl -s https://www.sqlite.org/download.html | grep "d391('a2'" | cut -d "'" -f 4`
This command gets the filename of the autoconf tarball from the SQLite official download page. See below.
![Screenshot 2023-03-24 at 14 10 51](https://user-images.githubusercontent.com/64626662/227429896-0e9d1366-11a3-48d1-90f3-b2b6a335689c.png)

However, that command has the potential to cause problems (e.g. HTML structure changed on the SQLite side).
So, it may be a way to create a PR like Dependabot does, so that the final decision is made by a human. :thinking:

**Reply by justinclift** [reply]
Good thinking! :smile:

The layout of the download page doesn't seem to change **too** much, so it seems like it should work.

We can try it out, and see how it goes. :smile:

Something along the lines of this should work:

```bash
TARBALL=$(curl -s https://www.sqlite.org/download.html | grep "d391('a2'" | cut -d "'" -f 4)
curl -LsS -o sqlite.tar.gz https://sqlite.org/${TARBALL}
```

I'll update our AppImage with that right now, as it's trivially easy to do.  If it seems to work ok, then @deepsidhu1313 can probably do the same + we can switch the AppImage to use @deepsidhu1313's build of SQLite.

**Reply by lucydodo** [reply]
How do you think this should be done in homebrew-sqlite3?
I don't think it would be possible to do the same within the 'Homebrew Formula', and I think the best way to do it would be to use GitHub Actions to check it at regular intervals and commit if there is a newer version.

**Reply by justinclift** [reply]
Yeah, that's going to be a bit trickier.  So far, the (manual) process I've been using (after noticing the new release) is:

1. On my desktop, `cd` into my local git clone of [our homebrew-sqlite3 tap](https://github.com/sqlitebrowser/homebrew-sqlite3)
   * The `Formula/sqlitefts5.rb` file in there is the one that needs changing
2. Update the `url`, `version`, and `sha256` fields with the new SQLite version
3. Comment out both `sha256 cellar:` lines in the `bottle` paragraph.  Lines 10 & 11.
4. Commit that change, and push to GitHub


5. Log into the appropriate macOS box (intel, arm) over ssh
6. Run "brew update" so it picks up the changed GitHub repo (above)
7. Uninstall all existing brew packages using:
   ```
   $ brew remove `brew list`
   ```
8. Build an updated Homebrew "bottle" for the new SQLite release:
   ```
   $ brew install --build-bottle sqlitefts5
   $ brew bottle sqlitefts5
   ```
9. Fix the filename generated by `brew bottle`, as it always slightly mucks it up.
   * eg: `mv sqlitefts5--3.41.1.arm64_monterey.bottle.1.tar.gz sqlitefts5-3.41.1.arm64_monterey.bottle.tar.gz`
10. rsync that file to the nightlies server:
   ```
   $ rsync -aP sqlitefts5-3.41.1.arm64_monterey.bottle.tar.gz nightlies:/nightlies/homebrew_bottles/
   ```

11. Log into the nightlies server and fix the ownership & permission of the uploaded file, so it can be downloaded via nginx.  (simple `chown` and `chmod`)
12. Repeat the above bottle building steps for the other macOS box

13. On my local desktop, uncomment the `bottle` sections of that `Formula/sqlitefts5.rb` file, update the sha256 values in it, commit that, and push to GitHub.

The next macOS nightly builds will automatically use this updated formula.

Looking over the above list, it's probably not super difficult to automate a large amount of it.  If we just do the first bit to update the SQLite tarball info, and we stop building bottles out of it... that should work fine.  It'll just add a few minutes to the build time for each macOS nightly build, as it compiles SQLite from tarball instead.  Very workable. :smile:

**Reply by justinclift** [reply]
Updating the AppImage build to use the new automatic approach worked:

![image](https://user-images.githubusercontent.com/406299/227441599-06eb3f1b-8e50-4bef-b1b9-dbb5cd9858d5.png)

**Reply by justinclift** [reply]
Heh, GitHub really likes to expand those images...

**Reply by justinclift** [reply]
The version of Qt shown in that screenshot is 5.9.x.  That's *ancient*.  Surprised it's not causing us problems.

Oh well. :shrug:

**Comment by justinclift** [comment]
Oh.  Looking over the SQLite download page, it already has a section for automated retrieval of version info:

```
For convenient, script-driven extraction of the downloadable file URLs and
associated information, an HTML comment is embedded in this page's source. Its first line (sans leading tag) reads:

    Download product data for scripts to read

Its subsequent lines comprise a CSV table with this column header ...
```

Looking at the source of the download page, it shows up in a HTML comment:

```
<!-- Download product data for scripts to read
PRODUCT,VERSION,RELATIVE-URL,SIZE-IN-BYTES,SHA3-HASH
PRODUCT,3.41.2,2023/sqlite-amalgamation-3410200.zip,2623078,c51ca72411b8453c64e0980be23bc9b9530bdc3ec1513e06fbf022ed0fd02463
PRODUCT,3.41.2,2023/sqlite-autoconf-3410200.tar.gz,3125545,1ebb5539dd6fde9a0f89e8ab765af0b9f02010fc6baf6985b54781a38c00020a
PRODUCT,3.41.2,2023/sqlite-doc-3410200.zip,10633275,8b3daa86d41cbc407e0992e11193d81094f01771c2dbeae93695f3dbaf19cbfe
PRODUCT,3.41.2,2023/sqlite-android-3410200.aar,3420407,6cd0ca943868978a945d4531bdc4e21fee8e3a7650e577f052550a37bc62db6e
PRODUCT,3.41.2,2023/sqlite-tools-linux-x86-3410200.zip,2269374,c41d2d97b62af6c6bd523ce9a06d1ea9f4e22e649f7a677a1c43f36d5ac3272c
PRODUCT,3.41.2,2023/sqlite-tools-osx-x86-3410200.zip,1616529,0e0f1eeade6ff7bc6f0ddbf0fe86798a595494328e1afa15da192e5fc516cb29
PRODUCT,3.41.2,2023/sqlite-dll-win32-x86-3410200.zip,576595,2a3b03595ee7c3b172465d9be3990fd8ac10fa93b9afd22031b67fa42c2325e1
PRODUCT,3.41.2,2023/sqlite-dll-win64-x64-3410200.zip,1202571,a85aa4c03f394147b4a9050245fb8d5d0b00ae726d1f259e71c07055e44854cf
PRODUCT,3.41.2,2023/sqlite-tools-win32-x86-3410200.zip,1999045,0ceebb7f8378707d6d6b0737ecdf2ba02253a3b44b1009400f86273719d98f1f
PRODUCT,3.41.2,2023/sqlite-wasm-3410200.zip,719097,1713c3a4c25b09f5d52ed25fb81948f6081e9eeefb5a6d449aead1cd848e962c
PRODUCT,3.41.2,2023/sqlite-src-3410200.zip,13836230,793e24c5158bafdb8a6e861ea9ad267cc773705d0bebb78b5c4aa323033e8028
PRODUCT,3.41.2,2023/sqlite-preprocessed-3410200.zip,2731397,6ffb450572964d829c688c3767448eac89f3577adfb1f08bfd3aaf0f050c3596
 -->
```

Looks like that should be "permanently stable".  And we could use the included checksum to verify the download.

Anyone want to figure out the magical curl/grep/cut/etc incantation to get that working? I'm too brain fried at the moment. :frowning:

**Reply by lucydodo** [reply]
Abracadabra! `curl -s https://sqlite.org/download.html | awk '/<!--/,/-->/ {print}' | grep 'autoconf'`
![Screenshot 2023-03-24 at 16 26 58](https://user-images.githubusercontent.com/64626662/227453304-eb579940-7c09-4431-80df-ad10bfa93f42.png)

**Reply by justinclift** [reply]
Thanks!  We can use `cut` on that to further extract just the field we want.

This seems to work:

```
curl -s https://sqlite.org/download.html | awk '/<!--/,/-->/ {print}' | grep 'sqlite-autoconf' | cut -d ',' -f 3
```

I'll update our AppImage commit from above with that instead. :smile:

**Reply by lucydodo** [reply]
The last hash value is SHA3-256. so we can't use that value.

**Reply by justinclift** [reply]
Hmmm, we should be able to.  Manually using the `openssl dgst` command to compute the sha3sum seems to work, matching the value recorded in the HTML:

```
$ openssl dgst -sha3-256 sqlite-autoconf-3410200.tar.gz
SHA3-256(sqlite-autoconf-3410200.tar.gz)= 1ebb5539dd6fde9a0f89e8ab765af0b9f02010fc6baf6985b54781a38c00020a
```

**Reply by justinclift** [reply]
This seems like it should work:

```
TARBALL=$(curl -s https://sqlite.org/download.html | awk '/<!--/,/-->/ {print}' | grep 'sqlite-autoconf' | cut -d ',' -f 3)
SHA3=$(curl -s https://sqlite.org/download.html | awk '/<!--/,/-->/ {print}' | grep 'sqlite-autoconf' | cut -d ',' -f 5)
curl -LsS -o sqlite.tar.gz https://sqlite.org/${TARBALL}
VERIFY=$(openssl dgst -sha3-256 sqlite.tar.gz | cut -d ' ' -f 2)
if [ "$SHA3" != "$VERIFY" ]; then exit 1 ; fi

tar -xzf sqlite.tar.gz
cd sqlite-autoconf-*
./configure --enable-shared=no
make
sudo make install
```

Anything look obviously wrong?

**Reply by lucydodo** [reply]
No. I was saying this because Homebrew Formulae uses SHA256 hashes. Never mind. :)

**Reply by lucydodo** [reply]
```bash
#!/bin/sh
export SQLITE_TARBALL=$(curl -s https://sqlite.org/download.html | awk '/<!--/,/-->/ {print}' | grep 'sqlite-autoconf' | cut -d ',' -f 3)
export SQLITE_TARBALL=https://sqlite.org/${SQLITE_TARBALL}
curl -LsS -o sqlite.tar.gz ${SQLITE_TARBALL}
export SHA256SUM=$(sha256sum sqlite.tar.gz | awk '{print $1}')
export VERSION=$(curl -s https://sqlite.org/download.html | awk '/<!--/,/-->/ {print}' | grep 'sqlite-autoconf' | cut -d ',' -f 2)

export CURRENT_VERSION=$(grep -o 'version "\S*"' sqlitefts5.rb | awk '{print $2}' | tr -d '"')
if [ "$VERSION" = "$CURRENT_VERSION" ]; then
    echo "It's already the latest version."
else
    sed -i '0,/url "/s|url "\(.*\)"|url "'${SQLITE_TARBALL}'"|' sqlitefts5.rb
    sed -i "s/version \"[^\"]*\"/version \"${VERSION}\"/" "sqlitefts5.rb"
    sed -i "s/sha256 \"[^\"]*\"/sha256 \"${SHA256SUM}\"/" "sqlitefts5.rb"
fi
```
Here is a shell script for the workflow to be written in homebrew-sqlite3.
I'd like to finalize and commit it now, but.. It's Friday evening and I have to leave for a beer with a friend. 😄 
Sorry, I'll finalize it tomorrow! Of course, contributions from others are always welcome.

**Reply by lucydodo** [reply]
Thanks for patience, I just created a [PR](https://github.com/sqlitebrowser/homebrew-sqlite3/pull/1).
